### PR TITLE
Refactor RayletBasedActorScheduler

### PR DESCRIPTION
**Pull request by Chong-Li**
## Why are these changes needed?

This is a starting PR for Ant's GCS-based actor scheduling.

In this PR:

- There is a new class 'RayletBasedActorScheduler', which is responsible for (randomly) selecting nodes for new actors and handling worker-lease-reply.
- The 'gcs_actor_schedule_strategy.h/.cc' are therefore removed.

In the next PR, we will introduce 'GcsBasedActorScheduler', which is comparable to 'RayletBasedActorScheduler'.

## Related issue number

#15943

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Chong-Li** [comment]
It seems that some rllib and serve test cases failed. These are not quite related to this GCS scheduling patch. It there any way to re-trigger CI without pushing new commits? @ffbin @rkooo567

**Comment by rkooo567** [comment]
They could be related because rllib + serve heavily uses actors. Btw, is this change not on by default now right? (that says, we are still using the old actor scheduling)?

**Comment by rkooo567** [comment]
For re-running CI, I can run Travis for you if you ask, and for buildkite, you should push a new commit unfortunately.

**Comment by Chong-Li** [comment]
> They could be related because rllib + serve heavily uses actors. Btw, is this change not on by default now right? (that says, we are still using the old actor scheduling)?

In this PR, we are just refactoring the old actor scheduling. The complete GCS-based one will be introduced in the next PR.

**Comment by Chong-Li** [comment]
> For re-running CI, I can run Travis for you if you ask, and for buildkite, you should push a new commit unfortunately.

Please do. I'd like to see if these failures are reproducible or not. Thx.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.h [hunk 1]
```diff
@@ -292,11 +304,42 @@ class GcsActorScheduler : public GcsActorSchedulerInterface {
...
   std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool_;
-  /// The actor schedule strategy.
-  std::shared_ptr<GcsActorScheduleStrategyInterface> actor_schedule_strategy_;
   /// The cached core worker clients which are used to communicate with leased worker.
   rpc::CoreWorkerClientPool core_worker_clients_;
 };
 
+/// RayletBasedActorScheduler randomly selects a node from the node pool for the new
+/// actor. If the worker leasing fails at the selected node, the corresponding Raylet
+/// would try to reply a spill-back node.
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.h:313]: Nit, better document what `RayletBasedActorScheduler` is here, and document the behavior of SelectNode and HandleWorkerLeaseReply on those methods.

**Review by raulchen — APPROVED** [review]

**Pull request merged**